### ISSUE TITLE
CellsUsed method does not consider data validation rules

### ISSUE THREAD

**Issue by Pankraty** [ClosedXML/ClosedXML#1026]
**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

0.93

**What is the current behavior?**

When the data validation rules are applied to cells these cells may be not presented in the `CellsUsed()` results if they have neither data, nor styles, nor conditional formats.

**What is the expected behavior or new feature?**

Complete this.

**Did this work in previous versions of our tool?  Which versions?**

No

**Code to reproduce problem:**
```c#
        [Test]
        public void ApplyingDataValidationMakesCellNotEmpty()
        {
            var wb = new XLWorkbook();
            var ws = wb.Worksheets.Add("Sheet1");
            ws.Range("B2:B12").SetDataValidation()
                .Decimal.EqualOrGreaterThan(0);

            var usedCells = ws.CellsUsed(includeFormats: true).ToList();

            Assert.AreEqual(11, usedCells.Count);
            Assert.AreEqual("B2", usedCells.First().Address.ToString());
            Assert.AreEqual("B12", usedCells.Last().Address.ToString());
        }
```
- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)

**Comment by igitur** [comment]
I think we should add a parameter `includeDataValidation` rather than rely on `includeFormats`. There are definitely some cases where I want to have cells return that have validation, but are empty otherwise.

**Comment by Pankraty** [comment]
I thought it would be too much of bool parameters so I've been re-writing the logic using a flagged enum. And gosh, there are _plenty_ of places where `includeFormats` is used...

**Comment by igitur** [comment]
I've been working on it too, but adding optional parameters and making `includeFormats` optional too. Yes, loads of places! Remember we have to be at least slightly backwards compatible.

**Comment by igitur** [comment]
In the end `XLCells.GetEnumerator()` will loop through the cells in `Worksheet.Internals.CellsCollection`. But empty cells with data validation do not get added to that collection (it seems at first glance).

**Comment by Pankraty** [comment]
Oh, seems we are doing the same job.

**Comment by DKR82** [comment]
> 
> 
> I've been working on it too, but adding optional parameters and making `includeFormats` optional too. Yes, loads of places! Remember we have to be at least **slightly backwards compatible**.

Hi there. I wonder if it is "by design" that CellsUsed() now returns "empty" cells with no inner text?

            XLWorkbook workbook;
            IXLWorksheet ws;

            string sampleSheet = sample.xlsx";
            workbook = new XLWorkbook(sampleSheet);
            ws = workbook.Worksheets.First();

            IXLCells cells = ws.Column(1).AsRange().CellsUsed();

            //Explicitly using the Contents option fixes the problem
            //IXLCells cells = ws.Column(0).AsRange().CellsUsed(XLCellsUsedOptions.Contents);

I am using merged cells, but only interested in getting the text of it. But the default option now seems to be "AllContents" which also includes "MergedCells". It´s no big deal to change our productive code to handle this (using the XLCellsUsedOption.Contents flag). 
I was just wondering if there was a technical neccesity for breaking the backwards compatibility here?

And please excuse my accidentally opened new issue #1145

**Comment by Pankraty** [comment]
Hi @DKR82 ,

In the beginning, there was a simple `CellsUsed` method that would return all cells that had been initialized - whether they have content, or their format differs, whatever. Then someone must have said, "Hey, I don't care about formats and simply want to get all cells with data". But if we want to copy everything from a worksheet to another one we definitely need to copy "empty" cells if they have custom formats. That's how the `includeFormats` flag seems to appear.

But then someone else faced the problem that if he has a workbook with a single conditional format covering a whole worksheet it may take an eternity to copy such worksheet as ClosedXML will try to instantiate billions of cells before dying from `OutOfMemoryException`. So this guy must have thought "OK, I don't want cells having only conditional formats and nothing else to be included in the collection - I rather copy conditional formats separately". And that's how we got an `includeConditionalFormats` flag.

Then yet someone else complained: "Hey! I got a worksheet that I use as a template which has merged cells combined in a very specific manner. Why not all of them are being copied when I try to create a new worksheet from this template?" ~And this is how we got `includeMergedRanges` flags.~ Kidding. To handle this situation we introduced a flagged enum that allows to tweak the behavior of `LastCellUsed` method (and all the others) for basically any purpose. Otherwise we would end up with a ton of boolean flags and method overloads.

To make life a bit easier, we prepared compound flags, such as `AllFormats` and `AllContents`. Where to put merged ranges was a tricky question. They are not just about styling because they affect cell contents too (e.g. when you merge cells `A1:C3` the contents of all cells except `A1` disappear - which does not happen when any styles or data validations rules apply). On the other hand, they are not just about contents as well. The final decision to put them into `AllContents` was more or less arbitrary. Was it a breaking change? I don't think so. For me it looks more as a bugfix because previously the method simply ignored merged cells no matter what flags were specified. Sorry that it broke you production code 🙇

_BTW, if you create a new worksheet in Excel, merge cells `F10:F11` and then press Ctrl+End you will get into `F11` - so Excel does not consider such cells as empty. If you apply data validation rules or conditional formats to empty cells Ctrl+End will not show them, so they are treated as empty_
